eos/utils/expression-parser_TEST.cc

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2021 Méril Reboud
+ *
+ * This file is part of the EOS project. EOS is free software;
+ * you can redistribute it and/or modify it under the terms of the GNU General
+ * Public License version 2, as published by the Free Software Foundation.
+ *
+ * EOS is distributed in the hope that it will be useful, but WITHOUT ANY
+ * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 59 Temple
+ * Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+#include <test/test.hh>
+
+#include <eos/utils/expression.hh>
+#include <eos/utils/expression-fwd.hh>
+#include <eos/utils/expression-maker.cc>
+#include <eos/utils/expression-parser-impl.hh>
+#include <eos/utils/expression-printer.cc>
+#include <eos/utils/kinematic.hh>
+#include <eos/utils/options.hh>
+#include <eos/utils/parameters.hh>
+
+#include <iostream>
+
+using namespace test;
+using namespace eos::exp;
+using namespace eos;
+
+
+class ExpressionEvaluator
+{
+    public:
+        double visit(BinaryExpression & e)
+        {
+            BinaryExpression::func f = BinaryExpression::Method(e._op);
+
+            return f(e._lhs.accept_returning<double>(*this), e._rhs.accept_returning<double>(*this));
+        }
+
+        double visit(ConstantExpression & e)
+        {
+            return e._value;
+        }
+
+        double visit(ObservableExpression & e)
+        {
+            return e._observable->evaluate();
+        }
+
+        double visit(ObservableNameExpression &)
+        {
+            throw InternalError("Encountered ObserableNameExpression in ExpressionEvaluator::visit");
+
+            return 0.0;
+        }
+};
+
+
+struct ExpressionTest
+{
+    const std::string _input;
+
+    Expression _e;
+    bool _completed;
+
+
+    typedef std::string::const_iterator It;
+    ExpressionParser<It> p;
+
+    ExpressionTest(const std::string input) :
+        _input(input)
+    {
+        It f(_input.begin()), l(_input.end());
+        _completed = qi::phrase_parse(f, l, p, ascii::space, _e);
+    }
+};
+
+
+class ExpressionParserTest :
+    public TestCase
+{
+public:
+    ExpressionParserTest() :
+    TestCase("expression_parser_test")
+    {
+    }
+
+    virtual void run() const
+    {
+        {
+            ExpressionTest test("1+2*3");
+
+            std::stringstream out;
+            ExpressionPrinter printer(out);
+            test._e.accept(printer);
+
+            ExpressionEvaluator evaluator;
+
+            TEST_CHECK(test._completed);
+            TEST_CHECK_EQUAL(test._e.accept_returning<double>(evaluator), 7);
```

Comment:
``7.0``

Reply:
In principle, this is parsed as integers, so the result is `(int) 7`. In practice, I am happy with `7.0` because `TEST_CHECK_EQUAL` doesn't test the type :)

Reply:
Ah no sorry, my mistake, the evaluator evaluates to `double`.

Reply:
Hm, but your visitor can only return a double!